### PR TITLE
✨ Add support for private dns zones

### LINKED ISSUE #1016: Implement Azure private dns for private clusters
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

https://docs.microsoft.com/en-us/azure/dns/private-dns-overview

/assign 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:  This PR adds private dns support for private clusters. When a cluster with an internal api server LB is created, capz will create a private dns zone with it and link it to the cluster's vnet. The internal LB IP is then added to the dns zone as a record with hostname `apiserver`. This allows reaching the API Server from within the vnet with FQDN `apiserver.$CLUSTER_NAME.capz.io`. It also fixes the [hairpin routing issue](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-troubleshoot#cause-4-accessing-the-internal-load-balancer-frontend-from-the-participating-load-balancer-backend-pool-vm) on all control plane VMs by adding a hosts entry so control planes resolve `apiserver` as localhost.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1016

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Add support for private dns zones
```

### CODE REVIEW

File: api/v1alpha3/types.go [hunk 1]
```diff
@@ -437,3 +437,9 @@ type SecurityProfile struct {
 	// set. Default is disabled.
 	EncryptionAtHost *bool `json:"encryptionAtHost,omitempty"`
 }
+
+// Record specifies a DNS record.
+type Record struct {
```
Comment (devigned) [inline api/v1alpha3/types.go:442]: Is this a specific type of DNS record, a cname, A record, etc?

File: cloud/types.go [hunk 2]
```diff
@@ -172,3 +172,12 @@ type TagsSpec struct {
 	Tags       infrav1.Tags
 	Annotation string
 }
+
+// PrivateZoneSpec defines the specification for a private DNS zone.
+type PrivateZoneSpec struct {
```
Comment (devigned) [inline cloud/types.go:177]: wdyt about `PrivateDnsZoneSpec`?

File: templates/test/cluster-template-prow-private.yaml [hunk 3]
```diff
@@ -126,9 +126,13 @@ spec:
...
     - - LABEL=etcd_disk
       - /var/lib/etcddisk
+    postKubeadmCommands:
+    - if [ -f /tmp/kubeadm-join-config.yaml ] || [ -f /run/kubeadm/kubeadm-join-config.yaml
+      ]; then echo '127.0.0.1   apiserver.${CLUSTER_NAME}.capz.io apiserver' >> /etc/hosts;
+      fi
     preKubeadmCommands:
-    - if [ -f /tmp/kubeadm.yaml ]; then iptables -t nat -A OUTPUT -p all -d ${AZURE_INTERNAL_LB_IP}
-      -j DNAT --to-destination 127.0.0.1; fi
+    - if [ -f /tmp/kubeadm.yaml ] || [ -f /run/kubeadm/kubeadm.yaml ]; then echo '127.0.0.1   apiserver.${CLUSTER_NAME}.capz.io
```
Comment (devigned) [inline templates/test/cluster-template-prow-private.yaml:134]: Nice!

**Review by devigned — COMMENTED** [review]
Had a couple questions. Looks great!

File: cloud/types.go (hunk 2 above)
Comment (CecileRobertMichon) [reply cloud/types.go]: changed to `PrivateDnsSpec` since this is more than the zone

File: api/v1alpha3/types.go (hunk 1 above)
Comment (CecileRobertMichon) [reply api/v1alpha3/types.go:442]: it's A or AAAA (mapping hostname to IP v4 or v6). Tried to clarify in comments and renamed name to hostname.

File: cloud/services/privatedns/privatedns.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+// New creates a new private dns service.
+func New(scope Scope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile on disk is currently no-op. OS disks should only be deleted and will create with the VM automatically.
```
Comment (nader-ziada) [inline cloud/services/privatedns/privatedns.go:53]: description needs update?

**Review by nader-ziada — COMMENTED** [review]
all good except for one small kind of nit

File: cloud/services/privatedns/privatedns.go (hunk 4 above)
Comment (CecileRobertMichon) [reply cloud/services/privatedns/privatedns.go]: good catch, fixed!

**Comment by voor** [comment]
Does Azure Private DNS allow setting a recursive resolver?  My very limited experience with it seemed to cause DNS issues, but there's a good chance it wasn't configured properly.

File: api/v1alpha3/types.go (hunk 1 above)
Comment (devigned) [reply api/v1alpha3/types.go:445]: I believe these types of records are specifically known as address records. Would you consider `AddressRecord`?

https://en.wikipedia.org/wiki/List_of_DNS_record_types

File: cloud/services/privatedns/privatedns_test.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package privatedns
+
+import (
```
Comment (devigned) [inline cloud/services/privatedns/privatedns_test.go:19]: Would you sort these imports?

**Review by devigned — CHANGES_REQUESTED** [review]
Couple requested changes. Per the sorted imports, I'll open a PR with a linter rule to enforce.

**Comment by devigned** [comment]
@jadarsie what do you think of private DNS entries with regard to Azure Stack?

**Comment by CecileRobertMichon** [comment]
> Does Azure Private DNS allow setting a recursive resolver? My very limited experience with it seemed to cause DNS issues, but there's a good chance it wasn't configured properly.

@voor it does not, but Azure VMs are already configured to have "a recursive DNS service that is provided separately as part of Azure's infrastructure" (source https://docs.microsoft.com/en-us/azure/dns/dns-domain-delegation#resolution-and-delegation). What kind of issues did you run into? I have not observed any issues while testing this so far.

File: cloud/services/privatedns/privatedns_test.go (hunk 5 above)
Comment (CecileRobertMichon) [reply cloud/services/privatedns/privatedns_test.go:19]: done

File: api/v1alpha3/types.go (hunk 1 above)
Comment (CecileRobertMichon) [reply api/v1alpha3/types.go]: good idea, done

**Review by devigned — APPROVED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
(testing something unrelated)

/test pull-cluster-api-provider-azure-capi-e2e

**Comment by CecileRobertMichon** [comment]
/test pull-cluster-api-provider-azure-capi-e2e

fixed the cred issue in #1043

**Comment by CecileRobertMichon** [comment]
different flake :( 

/retest

**Comment by dlipovetsky** [comment]
> It also fixes the [hairpin routing issue](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-troubleshoot#cause-4-accessing-the-internal-load-balancer-frontend-from-the-participating-load-balancer-backend-pool-vm) on all control plane VMs by adding a hosts entry so control planes resolve `apiserver` as localhost.

For reference, the page linked in the description doesn't include the relevant issue; I think it was moved here: https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-troubleshoot-backend-traffic#cause-4-accessing-the-internal-load-balancer-frontend-from-the-participating-load-balancer-backend-pool-vm

**Pull request merged**